torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,375 @@
+
+    def last_parametrization(self):
+        # TODO Write docs
+        last = self
+        while last.is_chained():
+            last = last.orig_param
+        return last
+
+    def chain(self, parametrization):
+        r""" It modifies the current parametrization, chaining a new one to it.
+        We have that ``p2.evaluate(p1.evaluate(t)) == p2.chain(p1).evaluate(t)``
+
+        The original parametrization is modified after chain is applied to it
+
+        Args:
+            parametrization (nn.Parametrization): the parametrization to be chained
+        Returns:
+            Module: The original module with the parametrization module appended to it
+        """
+        if not isinstance(parametrization, Parametrization):
+            raise ValueError("The object '{}' is not a Parametrization"
+                             .format(parametrization))
+
+        if self.is_registered() and parametrization.is_registered():
+            raise ValueError("Both parametrizations are registered to a tensor. "
+                             "The chaining is ambiguous.")
+        last = self.last_parametrization()
+        is_parametrized = hasattr(last, "orig")
+        if is_parametrized:
+            # We move the orig tensor to the new last parametrization
+            orig = last.orig
+            delattr(last, "orig")
+
+        _set_parametrization(last, parametrization, "orig", set_cache=False)
+        if is_parametrized:
+            new_last = parametrization.last_parametrization()
+            if isinstance(orig, Parameter):
+                new_last.register_parameter("orig", orig)
+            else:
+                new_last.register_buffer("orig", orig)
+        return self
+
+    def is_chained(self):
+        # TODO Write docs
+        return is_parametrized(self, "orig")
+
+    def is_registered(self):
+        # TODO Write docs
+        try:
+            self.original()
+            return True
+        except ValueError:
+            return False
+
+
+def _set_parametrization(module, parametrization, tensor_name, set_cache):
+    r""" Sets up the parametrization mechanism used by parametrizations.
+    This works by substituting the class of the module by a class
+    that extends it and makes `tensor_name` into a property. It also
+    regsters the parametrization under `tensor_name + "_param"`.
+    If ``set_cache == True``, it also sets up the caching mechanism
+    """
+
+    # Define the getter and set the caching mechanism
+    if set_cache:
+        def get_parametrized(module):
+            cache = getattr(module, tensor_name + "_cache")
+            if cache is None:
+                param = getattr(module, tensor_name + "_param")
+                return param.evaluate()
+            else:
+                return cache
+    else:
+        # Simplified version of the above
+        def get_parametrized(module):
+            param = getattr(module, tensor_name + "_param")
```

Comment:
I'm not so sure about the "append a prefix to the attribute" convention. Maybe it would be nicer to add something a'la `ModuleDict` called `.parametrizations` and then you'd be able to do `linear.parametrizations.weight`?

Reply:
I took this convention from the current implementation of the pruning modules and `spectral_norm` and `weight_norm`. The problem with adding a new module dict is that one has to implement all the related stuff to iterate over it, and make it appear in `__repr__` and so on, which is quite painful if this is implemented outside of Module. Maybe changing it to `tensor_name + "_parametrization"`?

Reply:
Sorry can you elaborate on the module dict issues? What's the problem with iterating over it? Also, parameters usually don't appear in `__repr__` so what's the deal there?

Reply:
Parametrizations are modules, and might have parameters and buffers of their own. As such, they should appear in `__repr__`, and they should be returned by methods like `children`, `modules`, so that `parameters` works correctly.

Reply:
I don't think that they necessarily have to appear in the `__repr__` so I wouldn't bother with that. About appearing in all the submodule-related methods then yes, they should appear there, but you get that for free if you use `ModuleDict`, so what's the problem? It's a lot simpler than doing string arithmetic

Reply:
Oh, I didn't know that those were iterated as sub-modules, but it does make a lot of sense. I just implemented it and works like a charm. It is also much cleaner.

I was thinking of doing the same and an `OrderedDict` for the caches. What do you think about that?